datastore/sync_entity.go

```diff
@@ -263,6 +293,164 @@ func (dynamo *Dynamo) InsertSyncEntitiesWithServerTags(entities []*SyncEntity) e
+	pkv := expression.Value(clientID)
+	keyCond := expression.KeyEqual(pkb, pkv)
+	exprs := expression.NewBuilder().WithKeyCondition(keyCond)
+	expr, err := exprs.Build()
+	if err != nil {
+		return syncEntities, fmt.Errorf("error building expression to get updates: %w", err)
+	}
+
+	input := &dynamodb.QueryInput{
+		ExpressionAttributeNames:  expr.Names(),
+		ExpressionAttributeValues: expr.Values(),
+		KeyConditionExpression:    expr.KeyCondition(),
+		FilterExpression:          expr.Filter(),
+		TableName:                 aws.String(Table),
+	}
+
+	out, err := dynamo.Query(input)
+	if err != nil {
+		return syncEntities, fmt.Errorf("error doing query to get updates: %w", err)
+	}
+	count := *out.Count
+
+	err = dynamodbattribute.UnmarshalListOfMaps(out.Items, &syncEntities)
+	if err != nil {
+		return syncEntities, fmt.Errorf("error unmarshalling updated sync entities: %w", err)
+	}
+
+	var i, j int64
+	for i = 0; i < count; i += maxTransactDeleteItemSize {
+		j = i + maxTransactDeleteItemSize
+		if j > count {
+			j = count
+		}
+
+		items := []*dynamodb.TransactWriteItem{}
+		for _, item := range syncEntities[i:j] {
+			if item.ClientID == item.ID && item.Deleted != nil && *item.Deleted {
+				continue
+			}
+
+			// Fail delete if race condition detected (modified time has changed).
+			if item.Version != nil {
+				cond := expression.Name("Mtime").Equal(expression.Value(*item.Mtime))
+				expr, err := expression.NewBuilder().WithCondition(cond).Build()
+				if err != nil {
+					return syncEntities, fmt.Errorf("error deleting sync entities for client %s: %w", clientID, err)
+				}
+
+				writeItem := dynamodb.TransactWriteItem{
+					Delete: &dynamodb.Delete{
+						ConditionExpression:       expr.Condition(),
+						ExpressionAttributeNames:  expr.Names(),
+						ExpressionAttributeValues: expr.Values(),
+						TableName:                 aws.String(Table),
+						Key: map[string]*dynamodb.AttributeValue{
+							pk: {
+								S: aws.String(item.ClientID),
+							},
+							sk: {
+								S: aws.String(item.ID),
+							},
+						},
+					},
+				}
+
+				items = append(items, &writeItem)
+			} else {
+				// If row doesn't hold Mtime, delete as usual.
+				writeItem := dynamodb.TransactWriteItem{
+					Delete: &dynamodb.Delete{
+						TableName: aws.String(Table),
+						Key: map[string]*dynamodb.AttributeValue{
+							pk: {
+								S: aws.String(item.ClientID),
+							},
+							sk: {
+								S: aws.String(item.ID),
+							},
+						},
+					},
+				}
+
+				items = append(items, &writeItem)
+			}
+
+		}
+
+		_, err = dynamo.TransactWriteItems(&dynamodb.TransactWriteItemsInput{TransactItems: items})
+		if err != nil {
+			return syncEntities, fmt.Errorf("error deleting sync entities for client %s: %w", clientID, err)
+		}
+	}
+
+	return syncEntities, nil
+}
+
+// IsSyncChainDisabled checks whether a given sync chain has been deleted
+func (dynamo *Dynamo) IsSyncChainDisabled(clientID string) (bool, error) {
```

Comment:
Again, it seems like we only need ClientID: clientID, ID: _someTagForDisabledMarker_, and just check if this kind of record exists or not, doesn't seem like we need the projection either.

Reply:
I see what you're saying here.  In my comment above I didn't realize you meant using the same value for id for all deleted chains.  I don't see a reason why we couldn't do it that way.  Is there a reason we shouldn't do it the way it is now?

Reply:
This is what I thought:
1) Something like DisabledMarkerItem rather than saving it as SyncEntity seems better to me, similar to what we have like `ServerClientUniqueTagItem`, we can just store what is really necessary and extend this struct with more fields specific to this kind of item.
2) Using a known fixed value for the ID seems better to me, since you can easily query this type of records using that (pk = clientID, sk = tag), using clientID directly seems like a waste of this field where we can have more info. And it follows the current convention like what we did for server/client tag items. We might have more kinds of special items in the future too. I don't see why this one have to specifically use clientID as its ID.